tests/unittests/test_lookback_window.py

```diff
@@ -0,0 +1,362 @@
+from tap_salesforce.salesforce import Salesforce
+import unittest
+from unittest import mock
+import tap_salesforce
+
+# mock 'Salesforce' class
+class MockSalesforce:
+    rest_requests_attempted = 0
+    jobs_completed = 0
+    login_timer = None
+    def __init__(self, *args, **kwargs):
+        return None
+
+    def login(self):
+        pass
+
+# mock args and return desired state, catalog and config file
+class MockParseArgs:
+    config = {}
+    discover = None
+    properties = None
+    state = None
+    def __init__(self, config, discover, properties, state):
+        self.config = config
+        self.discover = discover
+        self.properties = properties
+        self.state = state
+
+# send args
+def get_args(config, discover=None, properties=None, state=None):
+    return MockParseArgs(config, discover, properties, state)
+
+class TestLookbackWindow(unittest.TestCase):
+    """
+        Test cases to verify the lookback window seconds are subtracted from the start date or state file date.
+    """
+
+    @mock.patch('tap_salesforce.Salesforce')
+    @mock.patch('singer.utils.parse_args')
+    def test_default_lookback_window(self, mocked_parse_args, mocked_Salesforce_class):
+        """
+            Test case to verify default lookback window (10 seconds) is passed if user has not passed from the config.
+        """
+
+        tap_salesforce.CONFIG = {
+            'refresh_token': None,
+            'client_id': None,
+            'client_secret': None,
+            'start_date': None
+        }
```

Comment:
As CONFIG is a global variable, and the tap is updating the variable [here](https://github.com/singer-io/tap-salesforce/blob/master/tap_salesforce/__init__.py#L374). So, when we will run the tests, the CONFIG will contain the previous values from the test (as shown in the image). Thus, we are resetting the value.

![image](https://user-images.githubusercontent.com/80324346/167794744-3a556d75-eff2-4dfa-a7ea-39e0f0b35c34.png)

Reply:
But you already mocked the value of `args.config` in the test
https://github.com/singer-io/tap-salesforce/blob/cc4eb605190f697c5330ef2b4dfa31782635e7fb/tests/unittests/test_lookback_window.py#L126
https://github.com/singer-io/tap-salesforce/blob/d89a18f4f4f95018cb508c8aea5522e74ea35b98/tap_salesforce/__init__.py#L373-L374

So I agree with what you are saying: the first test will have the `CONFIG` with `None`s in the values and every test after that will have previous values.

But in this case it doesn't matter because your test is mocking the value before `tap_salesforce.main_impl()` is called. So to reset the config to `{'refresh_token': None, ...}` doesn't need to happen since you immediately reset it to `get_args()`

---

This is what the code looks like now

```
tap-salesforce config   {'refresh_token': None, ...}

           First test   SET {'refresh_token': None, ...}
           First test   SET {'refresh_token': 'test_refresh_token', ...}
           First test   RUN test

          Second test   SET {'refresh_token': None, ...}
          Second test   SET {'refresh_token': 'test_refresh_token', ...}
          Second test   RUN test
```


I'm saying these are redundant

```
tap-salesforce config   {'refresh_token': None, ...}

           First test   SET {'refresh_token': None, ...}   # redundant
           First test   SET {'refresh_token': 'test_refresh_token', ...}
           First test   RUN test

          Second test   SET {'refresh_token': None, ...}   # redundant
          Second test   SET {'refresh_token': 'test_refresh_token', ...}
          Second test   RUN test

```

So we can simplify the test to just have

```
tap-salesforce config   {'refresh_token': None, ...}

           First test   SET {'refresh_token': 'test_refresh_token', ...}
           First test   RUN test

          Second test   SET {'refresh_token': 'test_refresh_token', ...}
          Second test   RUN test
```

Reply:
Got it. Removed the code to reset `CONFIG` in the unittest.